src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs

```diff
@@ -179,14 +180,38 @@ void WriteToChannel(ResourceEvent resourceEvent) =>
     /// <param name="resource">The resource to update</param>
     /// <param name="resourceId"> The id of the resource.</param>
     /// <param name="stateFactory">A factory that creates the new state based on the previous state.</param>
-    public Task PublishUpdateAsync(IResource resource, string resourceId, Func<CustomResourceSnapshot, CustomResourceSnapshot> stateFactory)
+    public async Task PublishUpdateAsync(IResource resource, string resourceId, Func<CustomResourceSnapshot, CustomResourceSnapshot> stateFactory)
     {
         var notificationState = GetResourceNotificationState(resource, resourceId);
 
-        lock (notificationState)
+        await notificationState.Lock.WaitAsync().ConfigureAwait(false);
+
+        try
         {
             var previousState = GetCurrentSnapshot(resource, notificationState);
 
+            // TODO: Order matters. Switch to KeyedCollection.
+            var commands = previousState.Commands.ToDictionary(c => c.Type);
+
+            foreach (var commandAnnotation in resource.Annotations.OfType<ResourceCommandAnnotation>())
+            {
+                var visible = true;
+                if (commandAnnotation.Visible is { } visibleFunc)
+                {
+                    visible = await visibleFunc(previousState).ConfigureAwait(false);
+                }
+                if (visible)
+                {
+                    commands[commandAnnotation.Type] = new ResourceCommandSnapshot(commandAnnotation.Type, commandAnnotation.DisplayName, commandAnnotation.IconContent, commandAnnotation.IsHighlighted);
+                }
+                else
+                {
+                    commands.Remove(commandAnnotation.Type);
+                }
+            }
```

Comment:
This doesn't look very efficient...

Reply:
It's run when a resource changes so it shouldn't run often. But I'm sure there are ways to be smarter here about changes.